### DISCUSSION TITLE
sb_frame.sp() <= active_frame.stack_pointer

### DISCUSSION THREAD

**Discussion by carlsverre** [SeaQL/FireDBG.for.Rust#38] [Bug]
A couple of my tests hit this issue when run. The weird thing is that not all of my tests do. Some work fine, some crash with this error. It's not super clear what the difference is between them.

Per #12 it sounds like this is related to inlining. In the below test I'm not doing any explicit inlining. Is there a way to configure cargo/rust/FireDBG to compile without inlining when running FireDBG? I would prefer to not disable it entirely or anything like that.

Here is an example of a failing test with some of the relevant details. PageRange is basically like core::ops::Range and ZERO_DIGEST is basically `[0;32]`. The various block methods are all trivial.

```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Block {
    pages: PageRange,
    digest: Digest,
}

    fn mk_block(start: usize, len: usize) -> Block {
        Block::new(start, len, ZERO_DIGEST.clone())
    }

    #[test]
    fn block_sanity() {
        let block = mk_block(0, 1);
        assert_eq!(block.start(), 0);
        assert_eq!(block.end(), 1);
        assert_eq!(block.pages(), &PageRange::new(0, 1));
    }
```

Here is the full backtrace when it fails:
```
running 1 test
thread 'main' panicked at debugger/src/debugger.rs:386:21:
assertion failed: sb_frame.sp() <= active_frame.stack_pointer
stack backtrace:
   0: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:144:5
   3: firedbg_rust_debugger::debugger::run::{{closure}}
   4: firedbg_rust_debugger::debugger::run
   5: firedbg_rust_debugger::debugger::Debugger::run
   6: firedbg_debugger::main::{{closure}}
   7: tokio::runtime::park::CachedParkThread::block_on
   8: firedbg_debugger::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

---

```
Running Arch Linux, all packages up to date as of Feb 16 2024.

Linux glimmer 6.7.4-arch1-1 #1 SMP PREEMPT_DYNAMIC Mon, 05 Feb 2024 22:07:49 +0000 x86_64 GNU/Linux
```

**Comment by tyt2y3** [comment]
I am not sure whether it's due to inlining. Do you have a small repo that can reproduce the issue?

You can turn on detailed logging by RUST_LOG=trace , might tell something.

**Comment by carlsverre** [comment]
I was able to produce a fairly minimal repro here: https://gist.github.com/carlsverre/f07c6c7fa6b3ea34023879f4317a40bc

Notably, I suspect the issue is related to Deref. Although not 100% sure.

**Reply by tyt2y3** [reply]
Thanks! I'll look into it.

**Reply by tyt2y3** [reply]
Strangely, I can't reproduce this, at least not on my M1.

**Reply by carlsverre** [reply]
That's frustrating. If it helps I'm using a Intel(R) Core(TM) i9-9900K CPU running Arch Linux and Linux version 6.7.6. Rust version 1.76

**Reply by carlsverre** [reply]
I added a full debug trace here: https://gist.github.com/carlsverre/f07c6c7fa6b3ea34023879f4317a40bc#file-debug_trace-out
